### PR TITLE
Support simplified event api for event only updates

### PR DESCRIPTION
**Pull request by aliabid94**
Partial fix for https://github.com/gradio-app/gradio/issues/7331

--------

Made several updates that make querying the API easier:

- Can listen to the updates of just one event id, instead of streaming updates of all events of a session_hash 
- A session hash is not required and will be given if not assigned. This allows users to make "sessionless" requests, so they don't have to worry about creating a session_hash for one off requests
- Support api_names for `queue/join`

See the following example for making a request via Python:

```python
import httpx
import json

evt = httpx.post('http://localhost:7860/queue/join/greet', json={
    "data": ["john"],
})
evt = dict(evt.json())

async with httpx.AsyncClient() as client:
    async with client.stream('GET', 'http://localhost:7860/queue/data', params=evt) as response:
        async for chunk in response.aiter_text():
            msg = json.loads(chunk[5:])
            if msg["msg"] == "process_completed":
                output = msg["output"]["data"]
                print(output)
```

I did not modify either client because
- the JS client is used by the gradio frontend, and we want that to listen to session_hashs to reduce the amount of connections, due to the browser limit
- all our backend tests use the python client, which should use the same routing as the JS client

### CODE REVIEW

**Comment by abidlabs** [comment]
Nice @aliabid94 I agree that these are nice simplifications but I wonder if we can simplify even further. Some thoughts:

1. This approach is not ideal for generators, since the messages returned by the `/queue/data` endpoint will look like:

"data": [[["append", [], "l"]]
"data": [[["append", [], "o"]]
"data": [[["append", [], "v"]]
...

2. Do we need both the `event_id` and the `session_hash` that is returned from `/queue/join`? Is `event_id` not enough?

3. The Python code you've written is nice, but it involves having to parse the returned data in a way that would be annoying to do in cURL / command line.

For those reasons^ instead of re-using the `/queue/join` and `/queue/data` endpoints, I'd suggest creating a separate endpoints for this. Say:

* POST to `/request/{api_name}` and in return, you get a unique URL that points to an [EventSource](https://developer.mozilla.org/en-US/docs/Web/API/EventSource) 
* GET the URL above and you get a stream of outputs -- just the outputs, nothing else to keep it dead simple to use with cURL / command line (ok fine maybe we include some metadata so we can manage errors)

Replicate does something like this fwiw: https://replicate.com/docs/streaming

**Comment by abidlabs** [comment]
> I will need to send heartbeats because the request may wait a while in queue, and there needs to be a message within a certain timeout or the connection will drop. So I will only send yielded output and heartbeats over this connections

Ok then maybe a simple protocol like:

{"status": "pending", result: ""}
{"status": "pending", result: ""}
{"status": "generating", result: "h"}
{"status": "generating", result: "e"}
{"status": "generating", result: "l"}
{"status": "generating", result: "l"}
{"status": "generating", result: "o"}
{"status": "complete", result: "hello"}

**Comment by abidlabs** [comment]
One other thing we should consider is how we want to handle files (i.e. any components that use the FileData type, such as gr.Image). Right now, the clients upload them to the Gradio app server and convert them to FileData objects before sending them to the prediction API. It would be a pain if a user had to manually provide a FileData dict using cURL / command line

**Comment by freddyaboulton** [comment]
Nice the approach @aliabid94 ! Agree with @abidlabs that we should not use the diffed output for generators in this new route. I think simplifying the returned fields like in @abidlabs 's comment [here](https://github.com/gradio-app/gradio/pull/7407#issuecomment-1942934580) would be useful but I think the `result`/`data` field should always be a list so that the API doesn't change depending on the number of output components. 

Also, I think we need `session_hash` for stateful demos like chatbots, no?

**Comment by aliabid94** [comment]
Ok redid this whole PR. New endpoints mimic replicate's structure:
- POST `/call/{api_name}`
upload data here, recieve event id
- GET `/call/{api_name}/{event_id}`
Get stream of events. Supported events: "generating", "complete", "error", "heartbeat"
This stream shows actual values in generating, not diffs

Also refactored all server messages to use pydantic structures, instead of throwing around jsons. Pretty important to typecheck what our API's return, and easier to refactor as well.

Example using curl:
Recording 2024-02-14 at 20 24 42

**Comment by abidlabs** [comment]
This looks awesome @aliabid94. Now that we have this simplified API, could we provide cURL examples in our View API page? Would be great for the visibility of this API:

image

**Review by pngwn — APPROVED** [review]
Looks great to me! Only thing to note is that it would be good to bring these classes across to the frontend as typescript interfaces.

File: gradio/utils.py [hunk 1]
```diff
@@ -1035,6 +1035,9 @@ def __setitem__(self, key: K, value: V) -> None:
             self.popitem(last=False)
         super().__setitem__(key, value)
 
+    def __getitem__(self, key: K) -> V:
```
Comment (abidlabs) [inline gradio/utils.py:1038]: Is there any reason you explicitly add this?

**Comment by abidlabs** [comment]
Works great, thanks @aliabid94! 

For anyone else who wants to test, these are the cURL commands:

* `curl -X POST -H "Content-Type: application/json" -d '{"data": ["Gradio", 4]}' http://127.0.0.1:7860/call/predict`
* `curl -X GET -H "Accept: text/event-stream" http://127.0.0.1:7860/call/predict/1897b393c005417e90104dda525d1701CC`

I'm excited to get this working with files and then surfacing this on the view API docs (but that can a separate PR)!

**Review by abidlabs — APPROVED** [review]

**Pull request merged**